### PR TITLE
Improvements to Tool and Component functionality

### PR DESCRIPTION
**Pull request by kosack**
This PR is part of an effort to make some standard tools for DL1/DL2 production in ctapipe.  One of the things I found lacking was the ability to see what the full configuration that was used in a Tool (i.e. the Tool's `config` attribute only gives you the options the user passed in, not the values used).  

To make it possible to generate a fully populated config dictionary for a Component or Tool, this adds a few features:

* Tools now have a `get_current_config()` method that returns the full config for an *instance* (this was only possibly for a *Class* with `traitlets.config.Application`).  This is automatically registered with the Provenance system during `Tool.run()`
* added an `export_tool_to_commented_yaml` function which takes a tool and its current config and writes out a YAML file with correct comments (it's a bit of a hack based on the `traitlets.Application.write_config()` code,, so perhaps not ready for production use, but it's a step in the direction of being able to read and write YAML configurations (so far only JSON and Python are supported for reading)
* Added some nice HTML representations of Tools and Components for Jupyter notebooks
* Tools must now "register" all sub-components so that their current configuration can be queried
* the functionality in `ctapipe.utils.tools` has now moved to `ctapipe.core.traits` (both packages had similar scope)

e.g.:

```py3
class MyTool:
    ...
    def setup(self):
        self.mycomp = self.add_component(MyComponent(parent=self))
```

Then when one wants to get the full config, it looks like this:

```py3
tool = DisplayIntegrator()
...
pprint(tool.get_current_config())
```

```
{
         "DisplayIntegrator": {
            "channel": 0,
            "config_file": "",
            "event_index": 0,
            "extractor_product": "NeighborPeakWindowSum",
            "log_datefmt": "%Y-%m-%d %H:%M:%S",
            "log_format": "%(levelname)s: %(message)s [%(name)s.%(funcName)s]",
            "log_level": 10,
            "telescope": 458,
            "use_event_id": false
         },
         "SimTelEventSource": {
            "allowed_tels": [],
            "input_url": "/Users/kkosack/Data/CTA/Prod3b/gamma/gamma_20deg_0deg_run100___cta-prod3_desert-2150m-Paranal-merged.simtel.gz",
            "max_events": null,
            "skip_calibration_events": true
         },
         "EventSeeker": {},
         "NeighborPeakWindowSum": {
            "lwt": 0,
            "window_shift": 3,
            "window_width": 7
         },
         "HESSIOR1Calibrator": {},
         "CameraDL0Reducer": {},
         "CameraDL1Calibrator": {
            "clip_amplitude": null,
            "radius": null
         }
      }
```

In making this update, I ran into a lot of confusion and undocumented features of `traitlets.config`, so it might be better later to re-implement `traitlets.config.Application` altogether (I started a small notebook to explore that, seen here: https://gist.github.com/kosack/8158a2748a458bbece74a010abb02d3c but found that it is too much effort to get all the functionality now.  Perhaps somebody else has the time)

### CODE REVIEW

**Comment by kosack** [comment]
For an example of the experimental YAML output:
```yaml
print(export_tool_config_to_commented_yaml(tool))                                                 
#------------------------------------------------------------------------------
# DisplayIntegrator(Tool) configuration
#------------------------------------------------------------------------------
DisplayIntegrator:
    # Channel to view
    # Choices: any of [0, 1]
    # Default: 0
    channel: 0

    # name of a configuration file with parameters to load in addition
    # to command-line parameters
    # Default: ''
    config_file: ""

    # Event index to view.
    # Default: 0
    event_index: 0

    # ImageExtractor to use.
    # Choices: any of ['FullWaveformSum', 'FixedWindowSum',
    # 'GlobalPeakWindowSum', 'LocalPeakWindowSum',
    # 'NeighborPeakWindowSum',
    # 'BaselineSubtractedNeighborPeakWindowSum'] or None
    # Default: 'NeighborPeakWindowSum'
    extractor_product: "NeighborPeakWindowSum"

    # The date format used by logging formatters for %(asctime)s
    # Default: '%Y-%m-%d %H:%M:%S'
    log_datefmt: "%Y-%m-%d %H:%M:%S"

    # The Logging format template
    # Default: '%(levelname)s [%(name)s] (%(module)s/%(funcName)s):
    # %(message)s'
    log_format: "%(levelname)s [%(name)s] (%(module)s/%(funcName)s): %(message)s"

    # Set the log level by value or name.
    # Choices: any of (0, 10, 20, 30, 40, 50, 'DEBUG', 'INFO', 'WARN',
    # 'ERROR', 'CRITICAL')
    # Default: 30
    log_level: 20

    # Telescope to view. Set to None to display the firsttelescope
    # with data.
    # Default: None
    telescope: None

    # event_index will obtain an event using event_id instead of
    # index.
    # Default: False
    use_event_id: False
```

File: ctapipe/tools/display_dl1.py [hunk 1]
```diff
@@ -1,31 +1,32 @@
...
 
 
 class ImagePlotter(Component):
+    """ Plotter for camera images """
+
     display = Bool(
-        True,
-        help='Display the photoelectron images on-screen as they '
-             'are produced.'
+        True, help="Display the photoelectron images on-screen as they " "are produced."
```
Comment (thomasarmstrong) [inline ctapipe/tools/display_dl1.py]: Extra "" or new line?

File: ctapipe/tools/display_dl1.py [hunk 2]
```diff
@@ -137,41 +142,31 @@ class DisplayDL1Calib(Tool):
     telescope = Int(
         None,
         allow_none=True,
-        help='Telescope to view. Set to None to display all '
-             'telescopes.'
+        help="Telescope to view. Set to None to display all " "telescopes.",
```
Comment (thomasarmstrong) [inline ctapipe/tools/display_dl1.py]: Extra ""

File: ctapipe/tools/display_dl1.py [hunk 3]
```diff
@@ -137,41 +142,31 @@ class DisplayDL1Calib(Tool):
...
             D=(
-                {
-                    'ImagePlotter': {
-                        'display': True
-                    }
-                },
-                "Display the photoelectron images on-screen as they "
-                "are produced."
+                {"ImagePlotter": {"display": True}},
+                "Display the photoelectron images on-screen as they " "are produced.",
```
Comment (thomasarmstrong) [inline ctapipe/tools/display_dl1.py]: Extra ""

Comment (kosack) [reply ctapipe/tools/display_dl1.py]: thanks- seems to be due to reformatting (strings like that get concatenated, but there's no need in this case)

File: ctapipe/core/tool.py [hunk 4]
```diff
@@ -101,8 +101,17 @@ def main():
...
-                                     "parameters to load in addition to "
-                                     "command-line parameters")).tag(config=True)
+                                    "parameters to load in addition to "
+                                    "command-line parameters")).tag(config=True)
+    log_format = Unicode(
+        '%(levelname)s [%(name)s] (%(module)s/%(funcName)s): %(message)s',
+        help='The Logging format template'
+    ).tag(config=True)
+
+    log_format = Unicode(
```
[the remark this answers is not in the document]
Comment (kosack) [reply ctapipe/core/tool.py]: Nope! I'll fix it

**Pull request merged**